tests/filtered_relation/tests.py

```diff
@@ -424,3 +585,182 @@ def test_aggregate(self):
+            rate=1.00
+        )
+        CurrencyConversion.objects.create(
+            date=cls.sales_date2,
+            from_currency=cls.currency1,
+            to_currency=cls.currency1,
+            rate=1.00
+        )
+        CurrencyConversion.objects.create(
+            date=cls.sales_date1,
+            from_currency=cls.currency2,
+            to_currency=cls.currency2,
+            rate=1.00
+        )
+        CurrencyConversion.objects.create(
+            date=cls.sales_date2,
+            from_currency=cls.currency2,
+            to_currency=cls.currency2,
+            rate=1.00
+        )
+
+        BookDailySales.objects.create(
+            book=cls.book1,
+            date=cls.sales_date1,
+            currency_code=cls.currency1,
+            sales=100.00,
+            seller=cls.seller1,
+        )
+        BookDailySales.objects.create(
+            book=cls.book2,
+            date=cls.sales_date1,
+            currency_code=cls.currency2,
+            sales=200.00,
+            seller=cls.seller1,
+        )
+        BookDailySales.objects.create(
+            book=cls.book1,
+            date=cls.sales_date2,
+            currency_code=cls.currency1,
+            sales=50.00,
+            seller=cls.seller2,
+        )
+        BookDailySales.objects.create(
+            book=cls.book2,
+            date=cls.sales_date2,
+            currency_code=cls.currency2,
+            sales=100.00,
+            seller=cls.seller2,
+        )
+
+    def test_with_summation_aggregation(self):
+        qs = Book.objects.annotate(
+            recent_sales=FilteredRelation(
+                'daily_sales',
+                condition=Q(daily_sales__date__gte='2020-07-07')
+            ),
+            recent_sales_conversions=FilteredRelation(
+                "recent_sales__currency_code__conversions_from",
+                condition=Q(
+                    recent_sales__currency_code__conversions_from__date=F("recent_sales__date"),
+                    recent_sales__currency_code__conversions_from__to_currency=self.currency1
+
+                )
+            )
+        ).annotate(
+            sales_sum=Sum(F("recent_sales__sales") * F("recent_sales_conversions__rate"), output_field=DecimalField())
+        ).values(
+            'title', 'sales_sum'
+        ).order_by(
+            F('sales_sum').desc(nulls_last=True), 'title'
+        )
+
+        self.assertSequenceEqual(qs, [
+            {
+                'title': self.book2.title, 'sales_sum': Decimal(200.00),
+            },
+            {
+                'title': self.book1.title, 'sales_sum': Decimal(50.00),
+            },
+            {
+                'title': self.book3.title, 'sales_sum': None,
+            },
+        ])
+
+    def test_with_summation_aggregation_and_negation(self):
+        qs = Book.objects.annotate(
+            recent_sales=FilteredRelation(
+                'daily_sales',
+                condition=~Q(daily_sales__seller=self.seller1)
```

Comment:
@charettes , this case is most similar to what you suggested we have tests cover, using negation on a nested relationship.

Reply:
This code path won't trigger `split_exclude` but from my local testing it seems `condition=~Q` already has a flawed implementation on `master` anyway.

```diff
     def test_with_exclude(self):
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).filter(book_alice__isnull=True),
+            [self.author2],
+        )
         self.assertSequenceEqual(
             Author.objects.annotate(
                 book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
             ).exclude(book_alice__isnull=False),
             [self.author2],
         )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).filter(book_alice__isnull=False),
+            [self.author1],
+        )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).exclude(book_alice__isnull=True),
+            [self.author1],
+        )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
+            ).filter(book_not_alice__isnull=False),
+            [self.author2, self.author2, self.author1],
+        )
+        # The following fails while all the other succeed.
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
+            ).exclude(book_not_alice__isnull=True),
+            [self.author2, self.author2, self.author1],
+        )

     def test_with_join_and_complex_condition(self):
         self.assertSequenceEqual(
```

Reply:
To summarize the issue mentioned above is not related to the patch here, looks like it already exists in master.

Reply:
Thanks for the `--debug-sql` tip, I wasn't aware of that. And thanks for those examples of what you mean. It looks like there was a failure in the `database=oracle18,label=bionic4,python=python3.8` tests and now after rebasing in the spelling checks in the docs, so I'll look into those tests today as well to see if I can get a better assertion in place.

Reply:
@charettes , Some other tests in this suite use a `distinct` to I think avoid the same issue, would you feel comfortable using a distinct to assert the right result? For example, in the last example that you posted, if you add a distinct, this passes:
```
        self.assertSequenceEqual(
            Author.objects.annotate(
                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
            ).filter(book_not_alice__isnull=False).order_by('id').distinct(),
            [self.author1, self.author2],
        )
```
with query:
```
SELECT DISTINCT "filtered_relation_author"."id", "filtered_relation_author"."name", "filtered_relation_author"."content_type_id", "filtered_relation_author"."object_id" FROM "filtered_relation_author" INNER JOIN "filtered_relation_book" book_not_alice ON ("filtered_relation_author"."id" = book_not_alice."author_id" AND (NOT (UPPER(book_not_alice."title"::text) = UPPER(poem by alice)))) WHERE book_not_alice."id" IS NOT NULL ORDER BY "filtered_relation_author"."id" ASC
```

I don't think it make sense to assert the `[self.author1, self.author2, self.author2]` result for the non-distinct one just to document the current behavior, we should probably try to push-down to a sub-query in that case, right? Do we have a separate ticket for the existing issue in `master`? There are some closed-fixed ones here that look related: https://code.djangoproject.com/search?q=split_exclude&noquickjump=1&changeset=on&milestone=on&ticket=on&wiki=on

Reply:
@ferrants I think this can be addressed in a separate ticket yes, the above diff was more a of demonstration of current issues with negated filtered conditions but you're probably right that it's related to existing tickets.